x-pack/plugins/observability_ai_assistant/public/components/chat/chat_body.tsx

```diff
@@ -410,3 +410,19 @@ export function ChatBody({
     </EuiFlexGroup>
   );
 }
+
+// Exported for testing only
+export function reverseToLastUserMessage(messages: Message[], message: Message) {
+  // Drop messages after and including the one marked for regeneration
+  const indexOf = messages.indexOf(message);
+  const previousMessages = messages.slice(0, indexOf);
+
+  // Go back to the last written user message to fully regenerate function calls
+  const lastUserMessageIndex = findLastIndex(
+    previousMessages,
+    (aMessage: Message) => aMessage.message.role === 'user' && !aMessage.message.name
```

Comment:
What does the check for message.name do?

Reply:
If you look at the test data, the flow looks something like this:

1. System message (role: system)
2. User written prompt (role: user)
3. Recall function call (role: assistant)
4. Recall function response (role: **user**)
5. LLM response (role: assistant)

So both the user written prompt **and** the function response have the user role so filtering only on that isn't enough, so by checking if there is a `name` property I'm able to ignore the function call response (since it has the name property set to the name of the function that was called)!